### ISSUE TITLE
possible bug: dhcp server is new style and dhcp client is old style

### ISSUE THREAD

**Issue by the-maldridge** [jeremmfr/terraform-provider-junos#313]
I'm still trying to track down if this is an actual bug or not, but right now it looks like if you configure a DHCP server using the local server hierarchy, this gets created in a way that's incompatible with a dhcp client created on a logical interface.

I'm still trying to work out what's going on here, but this blog post is what led me to believe this is the case: https://networkshinobi.wordpress.com/2016/04/24/juniper-srx-new-dhcp-configuration-for-home-lan/

A second opinion on what's going on would be really appreciated, right now during apply I get the following error:

```
  # junos_interface_logical.upstream will be updated in-place
  ~ resource "junos_interface_logical" "upstream" {
      + description                = "Upstream Copper Handoff"
        id                         = "ge-0/0/0.0"
        name                       = "ge-0/0/0.0"
      + routing_instance           = "core"
        # (4 unchanged attributes hidden)

      ~ family_inet {
            # (3 unchanged attributes hidden)

          ~ dhcp {
              ~ update_server                                  = false -> true
              + vendor_id                                      = "juniper-srx300"
                # (10 unchanged attributes hidden)
            }
        }
    }

Plan: 0 to add, 7 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

junos_interface_logical.upstream: Modifying... [id=ge-0/0/0.0]
junos_interface_logical.upstream: Still modifying... [id=ge-0/0/0.0, 10s elapsed]
╷
│ Error: failed to netconf commit : netconf rpc [error]
│ [edit interfaces ge-0/0/0 unit 0 family inet]
│   'dhcp'
│     mgd: Incompatible with the client configured under 'family inet dhcp-client'
│ 
│   with junos_interface_logical.upstream,
│   on upstream.tf line 10, in resource "junos_interface_logical" "upstream":
│   10: resource "junos_interface_logical" "upstream" {
│ 
╵

```

**Comment by jeremmfr** [comment]
With SRX devices and old version of Junos, there is two way for DHCP configuration
- a legacy DHCP configuration with : 
```
# for server 
edit system services dhcp
# for client
edit interfaces <interface-name> unit <logical-unit-number> family inet dhcp
```
- an extended DHCP configuration with :
```
# for server 
edit system services dhcp-local-server
edit access address-assignment pool
# for client
edit interfaces <interface-name> unit <logical-unit-number> family inet dhcp-client
```

But as explained [here](https://www.juniper.net/documentation/us/en/software/junos/dhcp/topics/ref/statement/dhcp-client-edit-interfaces-srx.html#dhcp-client__d31055e136) and [here](https://www.juniper.net/documentation/us/en/software/junos/dhcp/topics/topic-map/dhcp-client-security-devices.html#id-example-configuring-the-device-as-a-dhcp-client), since Junos OS Release 15.1X49-D60 and Junos OS Release 17.3R1, the legacy configuration is deprecated and the client configuration in path
```
edit interfaces <interface-name> unit <logical-unit-number> family inet dhcp-client
```
has been moved to
```
edit interfaces <interface-name> unit <logical-unit-number> family inet dhcp
```
and `dhcp-client` no longer exists.

So I think you have an old version of Junos.
I will take care of adding an option to be able to configure the dhcp client with the old `dhcp-client` path.

**Comment by the-maldridge** [comment]
Absolutely correct on the old version, I'm on `15.1X49-D230` on this hardware in particular, and I have a few SRX220s that are even further behind.  This puzzles me though because I am on a later version, and so its not entirely clear why I'm not able to roll forward with the extended syntax.  It seems like I get an error that its incompatible in some way.  I'm using the new style dhcp local server group configuration as generated by terraform, so I'm relatively convinced its not that.

**Comment by jeremmfr** [comment]
I confirm that the client configuration `dhcp-client` has been moved to `dhcp` in recent releases.  
Example with a vSRX and version 18.3R1.9 :

```junos
[edit interfaces ge-0/0/3 unit 0 family inet]
user@vSRX-addr-0# set ?
Possible completions:
> accounting           Configure interface-based accounting options
> address              Interface address/destination prefix
+ apply-groups         Groups from which to inherit configuration data
+ apply-groups-except  Don't inherit configuration data from these groups
  arp-max-cache        Max interface ARP nexthop cache size (1..2000000)
  arp-new-hold-limit   Max no. of new unresolved nexthops (1..2000000)
> dhcp                 Dynamic Host Configuration Protocol client configuration
> filter               Packet filtering
  mtu                  Protocol family maximum transmission unit
  no-neighbor-learn    Disable neighbor address learning on interface
  no-redirects         Do not redirect traffic
> policer              Interface policing
  primary              Candidate for primary interface in system
> rpf-check            Enable reverse-path-forwarding checks on this interface
> sampling             Interface sampling
> simple-filter        Filter for doing multifield classification
> targeted-broadcast   Directed broadcast
  unconditional-src-learn  Glean from arp packets even when source cannot be validated
> unnumbered-address   Unnumbered interface address/destination
```

and if I try to use the old configuration name `dhcp-client`, the cli replace automatically the config :

```junos
[edit interfaces ge-0/0/3 unit 0 family inet]
user@vSRX-addr-0# show

[edit interfaces ge-0/0/3 unit 0 family inet]
user@vSRX-addr-0# set dhcp-client

[edit interfaces ge-0/0/3 unit 0 family inet]
user@vSRX-addr-0# show
dhcp;

```

But I think the version 15.1X49-D60 specified in Juniper documentation [here](https://www.juniper.net/documentation/us/en/software/junos/dhcp/topics/ref/statement/dhcp-client-edit-interfaces-srx.html#dhcp-client__d31055e136) and [here](https://www.juniper.net/documentation/us/en/software/junos/dhcp/topics/topic-map/dhcp-client-security-devices.html#id-example-configuring-the-device-as-a-dhcp-client) is not the correct one.  
It's maybe only after 17.3R1 or 17.4R1 (it's different depending on the doc page).

I launch a vSRX with version 15.1X49-D172.1 (newer than 15.1X49-D60) and `dhcp-client` has not yet been replaced by `dhcp` :

```junos
[edit interfaces ge-0/0/3 unit 0 family inet]
user@vSRX# set ?
Possible completions:
> accounting           Configure interface-based accounting options
> address              Interface address/destination prefix
+ apply-groups         Groups from which to inherit configuration data
+ apply-groups-except  Don't inherit configuration data from these groups
> dhcp-client          Dynamic Host Configuration Protocol client configuration
> filter               Packet filtering
  mtu                  Protocol family maximum transmission unit
  no-neighbor-learn    Disable neighbor address learning on interface
  no-redirects         Do not redirect traffic
> policer              Interface policing
  primary              Candidate for primary interface in system
> rpf-check            Enable reverse-path-forwarding checks on this interface
> sampling             Interface sampling
> simple-filter        Filter for doing multifield classification
> targeted-broadcast   Directed broadcast
  unconditional-src-learn  Glean from arp packets even when source cannot be validated
> unnumbered-address   Unnumbered interface address/destination prefix
```

moreover, there is always the `dhcp` config **hidden** for legacy DHCP configuration :

```junos
[edit interfaces ge-0/0/3 unit 0 family inet]
user@vSRX# set dhcp ?
Possible completions:
  <[Enter]>            Execute this command
+ apply-groups         Groups from which to inherit configuration data
+ apply-groups-except  Don't inherit configuration data from these groups
> client-identifier    DHCP Server identifies a client by client-identifier value
  lease-time           Lease time in seconds requested in DHCP client protocol packet (60..2147483647 seconds)
  retransmission-attempt  Number of attempts to retransmit the DHCP client protocol packet (0..6)
  retransmission-interval  Number of seconds between successive retransmission (4..64 seconds)
  server-address       DHCP Server-address
  update-server        Propagate TCP/IP settings to DHCP server
  vendor-id            Vendor class id for the DHCP Client
  |                    Pipe through a command
```

To sum up, the config name `dhcp` on SRX is for :

- **recent versions**: the extended DHCP configuration and there isn't a way to use the legacy configuration
- **older versions**: the legacy DHCP configuration.  

and the config name `dhcp-client` (for the extended DHCP configuration) is only present on the older versions.  

So, the provider uses the name `dhcp` when applying resource (to be the extended DHCP configuration on recent versions), and in your case (with older version) it's the legacy DHCP configuration and yes it's incompatible with other dhcp resources made for extended DHCP configuration.  

I confirm my plan to add an argument to be able to use the old name `dhcp-client` instead of `dhcp` in the case of an old version Junos on SRX.
